web/src/lib/components/memory-page/memory-viewer.svelte

```diff
@@ -90,32 +90,39 @@
 		}
 	};
 
-	let autoPlayInterval: NodeJS.Timeout;
 	let autoPlay = false;
 	let autoPlaySpeed = 5000;
 	let autoPlayProgress = 0;
 	let autoPlayIndex = 0;
-	let canPlayNext = true;
+	let autoPlayAnimationFrameRequest: number;
 
 	const toggleAutoPlay = () => {
 		autoPlay = !autoPlay;
-		if (autoPlay) {
-			autoPlayInterval = setInterval(() => {
-				if (!canPlayNext) return;
-
-				window.requestAnimationFrame(() => {
-					autoPlayProgress++;
-				});
-
-				if (autoPlayProgress > 100) {
-					autoPlayProgress = 0;
-					canPlayNext = false;
-					autoPlayTransition();
-				}
-			}, autoPlaySpeed / 100);
-		} else {
-			clearInterval(autoPlayInterval);
+		if (!autoPlay) {
+			cancelAnimationFrame(autoPlayAnimationFrameRequest);
+			return;
 		}
+
+		let previousTransition;
+		function draw(now: number) {
+			previousTransition ??= now;
+
+			const elapsed = previousTransition - now;
+			const progress = (elapsed / autoPlaySpeed) * 100;
+
+			if (progress >= 100) {
+				autoPlayProgress = 0;
+				autoPlayTransition();
+				previousTransition = now;
+			} else {
+				autoPlayProgress = progress;
+			}
+
+			if (autoPlay) {
+				autoPlayAnimationFrameRequest = requestAnimationFrame(draw);
+			}
```

Comment:
This condition might not be necessary given the animation frame is explicitly cancelled when autoplay is disabled.